datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java

```diff
@@ -406,28 +410,40 @@ public UpdateResponse update(String key, Datastream datastream) {
 
     if (!DatastreamStatus.READY.equals(datastream.getStatus()) && !DatastreamStatus.PAUSED.equals(datastream.getStatus())) {
       _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_405_METHOD_NOT_ALLOWED,
-          "Can only pause a datastream in READY/PAUSED state: " + datastreamName);
+          "Can only stop a datastream in READY/PAUSED state: " + datastreamName);
     }
 
     List<Datastream> datastreamsToStop =
         force ? getGroupedDatastreams(datastream) : Collections.singletonList(datastream);
     LOG.info("Stop datastreams {}", datastreamsToStop);
     for (Datastream d : datastreamsToStop) {
       try {
-        if (DatastreamStatus.READY.equals(datastream.getStatus()) || DatastreamStatus.PAUSED.equals(datastream.getStatus())) {
-          d.setStatus(DatastreamStatus.STOPPED);
+        if (DatastreamStatus.READY.equals(d.getStatus()) || DatastreamStatus.PAUSED.equals(d.getStatus())) {
+          d.setStatus(DatastreamStatus.STOPPING);
+          _store.updateDatastream(d.getName(), d, true);
+          _store.deleteDatastreamNumTasks(d.getName());
+        } else if (DatastreamStatus.STOPPING.equals(d.getStatus())) {
+          // this check helps in preventing any datastream from being stuck in STOPPING state indefinitely
+          LOG.warn("Datastream {} is already in {} state. Notifying leader to initiate transition", d,
+              d.getStatus());
           _store.updateDatastream(d.getName(), d, true);
           _store.deleteDatastreamNumTasks(d.getName());
         } else {
-          LOG.warn("Cannot stop datastream {}, as it is not in READY/PAUSED state. State: {}", d, datastream.getStatus());
+          LOG.warn("Cannot stop datastream {}, as it is not in READY/PAUSED state. State: {}", d, d.getStatus());
         }
       } catch (DatastreamException e) {
         _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_500_INTERNAL_SERVER_ERROR,
-            "Could not update datastream to STOPPED state: " + d.getName(), e);
+            "Could not update datastream to STOPPING state: " + d.getName(), e);
       }
     }
 
-    LOG.info("Completed request for stopping datastream {}", datastream);
+    // polls until the leader transitions the state of the datastream to STOPPED state
+    PollUtils.poll(() -> datastreamsToStop.stream()
+            .allMatch(ds -> _store.getDatastream(ds.getName()).getStatus().equals(DatastreamStatus.STOPPED)),
+        allStopped -> allStopped, TRANSITION_RETRY_PERIOD_MS_DEFAULT.toMillis(),
+        TRANSITION_TIMEOUT_MS_DEFAULT.toMillis()).orElseThrow(RetriesExhaustedException::new);
```

Comment:
What will be the Status code returned after the timeout? Can you define a deterministic behavior and return appropriate error message and error code?